mlptrain/sampling/md.py

```diff
@@ -282,37 +294,104 @@ def save_trajectory():
 
-        dyn.run(steps=n_steps)
+        if 'constraints' in kwargs and kwargs['constraints'] is not None:
+            ase_atoms.set_constraint(kwargs['constraints'])
 
-    traj = _convert_ase_traj(traj_name)
+        return mlp.ase_calculator
 
-    for i, (frame, energy) in enumerate(zip(traj, energies)):
-        frame.update_attr_from(configuration)
-        frame.energy.predicted = energy
-        frame.time = dt * interval * i
 
-    return traj
+def _run_dynamics(ase_atoms, ase_traj, traj_name, interval, temp, dt, dt_ase,
+                  n_steps, energies, biased_energies, calculator, bias,
+                  **kwargs) -> None:
+    """Initialise dynamics object and run dynamics"""
+
+    if temp > 0:                                        # Default Langevin NVT
+        dyn = Langevin(ase_atoms, dt_ase,
+                       temperature_K=temp,
+                       friction=0.02)
+    else:                                               # Otherwise NVE
+        dyn = VelocityVerlet(ase_atoms, dt_ase)
+
+    def append_unbiased_energy():
+        _append_unbiased_energy(ase_atoms, energies, calculator, bias)
+
+    def append_biased_energy():
+        _append_biased_energy(ase_atoms, biased_energies, bias)
+
+    def save_trajectory():
+        _save_trajectory(ase_traj, traj_name, **kwargs)
+
+    dyn.attach(append_unbiased_energy, interval=interval)
+    dyn.attach(append_biased_energy, interval=interval)
+    dyn.attach(ase_traj.write, interval=interval)
+
+    if any(key in kwargs for key in ['save_fs', 'save_ps', 'save_ns']):
+        dyn.attach(save_trajectory,
+                   interval=_traj_saving_interval(dt, kwargs))
+
+    logger.info(f'Running {n_steps:.0f} steps with a timestep of {dt} fs')
+    dyn.run(steps=n_steps)
+
+    # The calling process waits until PLUMED process has finished
+    if isinstance(bias, PlumedBias):
+        calculator.plumed.finalize()
+
+    return None
+
 
+def _append_unbiased_energy(ase_atoms, energies, calculator, bias) -> None:
+    """Append unbiased energy (biased MLP energy - bias energy) to the
+    trajectory"""
 
-def _save_trajectory(ase_traj: 'ase.io.trajectory.Trajectory',
-                     traj_name: str,
-                     **kwargs
-                     ) -> None:
-    """Saves the trajectory with a unique name based on the current simulation
+    if isinstance(bias, PlumedBias):
+        energy = calculator.calc.get_potential_energy(ase_atoms)
+
+    else:
+        energy = calculator.get_potential_energy(ase_atoms)
+
+    energies.append(energy)
+    return None
+
+
+def _append_biased_energy(ase_atoms, biased_energies, bias) -> None:
+    """Append bias energy to the trajectory"""
+
+    biased_energy = ase_atoms.get_potential_energy()
+
+    # Plumed calculator when get_potential_energy() is called returns an array
+    if isinstance(bias, PlumedBias):
+        biased_energy = biased_energy[0]
+
+    biased_energies.append(biased_energy)
```

Comment:
is this better? maybe

```suggestion
    result = ase_atoms.get_potential_energy()

    # Plumed calculator when get_potential_energy() is called returns an array
    biased_energies.append(result[0] if isinstance(bias, PlumedBias) else result)
```

Reply:
would be better if `get_potential_energy` returned something consistent! if this is an ASE problem I would consider subclassing the calculator and overriding `get_potential_energy`